Add canonical mount options zfs-mount-generator

aerusso opened the pull request:
`lib/libzfs/libzfs_mount.c:zfs_add_options` provides the canonical mount options used by a `zfs mount` command. Because we cannot call `zfs mount` directly from a systemd.mount unit, we mirror that logic in zfs-mount-generator.

The zed script is updated to cache these properties as well.

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Describe your changes in detail -->
#7449 clarifies that the short-term resolution for mounting via the `mount -t zfs -o zfsutil` approach will need to pass all mount options through `-o`. This patch, by referring to `lib/libzfs/libzfs_mount.c:zfs_add_options`, ensures these options are used when the zfs-mount-genererator mechanism is used to mount the datasets.

### How Has This Been Tested?
I've booted with these changes, and it resolves the proximate issue that read-only datasets are mounted ("temporary") rw.
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

aerusso commented on cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in:
```diff
@@ -44,9 +44,10 @@ case "${ZEVENT_HISTORY_INTERNAL_NAME}" in
-            canmount=*|mountpoint=*) ;;
+        # Only act if one of the tracked properties is altered.
+        case "${ZEVENT_HISTORY_INTERNAL_STR%%=*}" in
+            canmount|mountpoint|atime|relatime|devices|exec|\
```
There's some formatting issue---do you mean in front of the backslash \\?

I checked, that seems to be OK.

rlaager replied:
Yes, that was supposed to be a backslash. If you can put a space in front of the backslash and indent the continuation line, that is the typical style.

aerusso commented on cmd/zed/zed.d/history_event-zfs-list-cacher.sh.in:
```diff
@@ -60,8 +61,8 @@ esac
 zed_lock zfs-list
 trap abort_alter EXIT
 
-"${ZFS}" list -H -tfilesystem -oname,mountpoint,canmount -r "${ZEVENT_POOL}" \
```
Sorry, confused about this one, too.

rlaager replied:
Yes, that was supposed to be a backslash. If you can put a space in front of the backslash and indent the continuation line, that is the typical style.

aerusso commented on etc/systemd/system-generators/zfs-mount-generator.in:
```diff
@@ -77,6 +78,79 @@ process_line() {
   # Escape the mountpoint per systemd policy.
   mountfile="$(systemd-escape "${2#?}").mount"
+  mountpoint="${2}"
+  dataset="${1}"
```
Agreed. Moreover, looking up these attributes by argument number (and later on with that mess of shifts) is opaque and messy. I pulled all of them into named variables to make this more readable.

aerusso commented on etc/systemd/system-generators/zfs-mount-generator.in:
```diff
@@ -97,10 +171,10 @@ After=zfs-import.target
+What=${dataset}
 Type=zfs
-Options=zfsutil,auto
+Options=defaults${opts},zfsutil
```
I replaced `/bin/mount` with a shim and logged its arguments when mounting a dataset. `defaults` appeared, first on that list. I'm just replicating that behavior. If it should be different there, we should open another issue because that seems like it might be very subtle.

rlaager replied:
If I understand you correctly, previously, "defaults" was showing up in the call to mount, but wasn't in the unit's Options line. If you continue to leave it out of the Options, does it still show up to mount?

aerusso replied:
No. I did something like

```sh
% mv /bin/mount /bin/mount.real
% cat > /bin/mount <<EOF
#!/bin/sh
echo "\$@" > /tmp/mount.log
/bin/mount.real "\$@"
EOF
% chmod +x /bin/mount
% zfs mount REDACTED/test
% cat /tmp/mount.log
--no-canonicalize -t zfs -o defaults,atime,relatime,dev,exec,rw,suid,nomand,zfsutil REDACTED/test /REDACTED/test
```
This exercises a call to `lib/libzfs/libzfs_mount.c:zfs_mount` coming from `cmd/zfs/zfs_main.c` (I believe). In any case, I want to mimic the behavior of calling `zfs mount` as closely as possible, hence the `defaults`.

rlaager replied:
Okay, I see the rationale for that change. It still seems to me that defaults is unnecessary, since that is supposed to be a no-op by definition. But it shouldn’t hurt.

bunder2015 commented:
nit: canonical

aerusso commented:
One of the "cosmetic" changes actually broke the zedlet. This revision fixes that, and should makes it backwards compatible with the current trunk version. Let's let this steep on my systems for another week, and I'll ping you again after it's had a bit more testing.

aerusso commented:
@AttilaFueloep @ccope Both of you have tried out #7329, would it be possible to test this update to that PR? It should be backwards-compatible, and a little diversity in testing would be helpful!

You can install this without rebuilding the whole package, just be sure to substitute the correct paths in `zfs-mount-generator` and `history_event-zfs-list-cacher.sh`

aerusso commented:
@AttilaFueloep You also suggested adding a command to simplify bootstrapping the cache file. Instead, I've added a tutorial to the manual page. I think this is superior, since it's really just a few commands that need to be run. I'd appreciate your feedback.

AttilaFueloep commented:
@aerusso Sorry, I'm short of time today and maybe tomorrow but I'll have a look ASAP.

aerusso commented:
@behlendorf I think this is probably ready. The only changes since the last round of reviews are a (very slight) modification of the build infrastructure in man8, and some improvement to the documentation for the zfs-mount-generator, most notably a very brief tutorial.

behlendorf commented:
@aerusso sounds good.  Merged.

AttilaFueloep commented:
@aerusso I did some basic testing and everything seems to work fine. Currently I'm travelling so I'll postpone more thorough full testing until I'm back. I'll report back if some problem arises.

Thanks for the man page addition, it will ease activating the generator.

My suggestion in the other PR was targeted at changes of the cache file format. The scenario is as follows. If an update of ZFS in an inactive BE introduces a cache fie format change, you'll end up with the new code and an old format cache file. So I'm not sure what happens if you boot that BE. If the code will handle the old format cache file gracefully everything is fine. If not I would've to remove the cache file on every update since I have no easy way to tell if ZFS will get updated and no way to generate the cache file in the correct format. (Hopefully this description is more comprehensible than my former one.)

aerusso commented:
@AttilaFueloep That makes sense---and I think having that code posted in that thread is a good idea. However, the versions that have actually been committed are compatible, so the only people who would be exposed to such upgrade issues (from experimental version) would already know to look at that thread. And, frankly, it's probably only the six people who've posted there who even looked at it.

AttilaFueloep commented:
@aerusso OK thanks, I'm fine with that and will keep an eye on cache file format changes introduced by possible future commits.
